packages/muelu/src/Graph/PairwiseAggregation/MueLu_NotayAggregationFactory_def.hpp

```diff
@@ -533,32 +602,40 @@ namespace MueLu {
 
       LO bestIdx = Teuchos::OrdinalTraits<LO>::invalid();
       magnitude_type best_mu = Teuchos::ScalarTraits<magnitude_type>::zero();
-      const magnitude_type aii     = Kokkos::ArithTraits<value_type>::abs(diagA_h(currentIdx));
-      const magnitude_type si      = Kokkos::ArithTraits<value_type>::abs(rowSum_h(currentIdx));
+      const magnitude_type aii     = Kokkos::ArithTraits<value_type>::real(diagA_h(currentIdx));
+      const magnitude_type si      = Kokkos::ArithTraits<value_type>::real(rowSum_h(currentIdx));
       for(auto entryIdx = row_map_h(currentIdx); entryIdx < row_map_h(currentIdx + 1); ++entryIdx) {
         const LO colIdx = static_cast<LO>(entries_h(entryIdx));
-        if(currentIdx == colIdx || localAggStat[colIdx] != READY || values_h(entryIdx) == KAT_zero) {
+        if(currentIdx == colIdx || localAggStat[colIdx] != READY || values_h(entryIdx) == KAT_zero || colIdx > numRows) {
           continue;
         }
 
-        const magnitude_type aij = Kokkos::ArithTraits<value_type>::abs(values_h(entryIdx));
-        const magnitude_type ajj = Kokkos::ArithTraits<value_type>::abs(diagA_h(colIdx));
-        const magnitude_type sj  = Kokkos::ArithTraits<value_type>::abs(rowSum_h(colIdx));
+        const magnitude_type aij = Kokkos::ArithTraits<value_type>::real(values_h(entryIdx));
+        const magnitude_type ajj = Kokkos::ArithTraits<value_type>::real(diagA_h(colIdx));
+        const magnitude_type sj  = - Kokkos::ArithTraits<value_type>::real(rowSum_h(colIdx)); // NOTE: The ghostedRowSum vector here has has the sign flipped from Notay's S
         if(aii - si + ajj - sj >= MT_zero) {
           const magnitude_type mu_top    = MT_two / ( MT_one/aii + MT_one/ajj );
           const magnitude_type mu_bottom = -aij + MT_one / (MT_one / (aii - si) + MT_one / (ajj - sj));
           const magnitude_type mu = mu_top / mu_bottom;
-          if(bestIdx == Teuchos::OrdinalTraits<LO>::invalid() || mu < best_mu) {
+
+          // Modification: Explicitly check the tie criterion here
+	  if (bestIdx == LO_INVALID ||  mu < best_mu * tie_less ||
+              (mu < best_mu*tie_more && orderingVector[colIdx] < orderingVector[bestIdx])) {
             best_mu = mu;
             bestIdx = colIdx;
+            *out << "[" << currentIdx << "] Column     UPDATED " << colIdx << ": "
```

Comment:
Do we really want the output here?

Reply:
That's "optional" output due to Luc's debugging sorcery.

If you don't set the debug variable, nothing prints :)